File: 2024/day12/Python/day12.py
from time import time
from collections import defaultdict

def calculate_region_properties(input_file):
    # Start the timer
    start_time = time()

    # Read and prepare grid data
    with open(input_file, "r") as file:
        grid_data = [line.strip() for line in file]
    rows, cols = len(grid_data), len(grid_data[0])

    # Initialize grid with a default value for out-of-bound checks
    grid = defaultdict(lambda: ".", {(r, c): value for r, row in enumerate(grid_data) for c, value in enumerate(row)})

    # Initialize results and visited set
    total_area_perimeter1 = 0
    total_area_perimeter2 = 0
    visited = set()

    # Process each cell
    for row in range(rows):
        for col in range(cols):
            if (row, col) in visited:
                continue

            # Initialize stack and region tracking
            stack = [(row, col)]
            visited.add((row, col))
            region = set(stack)

            # Explore the region using DFS
            while stack:
                current_row, current_col = stack.pop()
                for neighbor_row, neighbor_col in [
                    (current_row + 1, current_col),
                    (current_row - 1, current_col),
                    (current_row, current_col + 1),
                    (current_row, current_col - 1),
                ]:
                    if (
                        grid[neighbor_row, neighbor_col] == grid[current_row, current_col]
                        and (neighbor_row, neighbor_col) not in visited
                    ):
                        stack.append((neighbor_row, neighbor_col))
                        visited.add((neighbor_row, neighbor_col))
                        region.add((neighbor_row, neighbor_col))

            # Calculate area and perimeters
            area = len(region)

            # Perimeter type 1: Edges not in the region
            perimeter1 = sum(
                1
                for r, c in region
                for nr, nc in [
                    (r + 1, c), (r - 1, c),
                    (r, c + 1), (r, c - 1)
                ]
                if (nr, nc) not in region
            )

            # Perimeter type 2: More complex condition
            perimeter2 = 0
            for r, c in region:
                if (r, c - 1) not in region and (
                    (r - 1, c) not in region or (r - 1, c - 1) in region
                ):
                    perimeter2 += 1
                if (r, c + 1) not in region and (
                    (r - 1, c) not in region or (r - 1, c + 1) in region
                ):
                    perimeter2 += 1
                if (r - 1, c) not in region and (
                    (r, c - 1) not in region or (r - 1, c - 1) in region
                ):
                    perimeter2 += 1
                if (r + 1, c) not in region and (
                    (r, c - 1) not in region or (r + 1, c - 1) in region
                ):
                    perimeter2 += 1

            # Accumulate results
            total_area_perimeter1 += area * perimeter1
            total_area_perimeter2 += area * perimeter2

    # Print results
    elapsed_time = time() - start_time
    print(f"Part 1 (Total Area x Perimeter 1): {total_area_perimeter1} ({elapsed_time:.3f}s)")
    print(f"Part 2 (Total Area x Perimeter 2): {total_area_perimeter2} ({elapsed_time:.3f}s)")

# Input file path
INPUT_FILE = "day12_input.txt"
calculate_region_properties(INPUT_FILE)

File: 2024/day12/R/day12.R
# Input file
INPUT_FILE <- 'day12_input.txt'
W <- nchar(readLines(INPUT_FILE)[1])
grid <- read.fwf(INPUT_FILE, widths = rep(1, W)) |> as.matrix()
rows <- nrow(grid)
cols <- W

# Initialize matrices
region_touching <- matrix(-1, rows, cols)
visited <- matrix(FALSE, rows, cols)

# Unique region values
region_values <- sort(unique(as.vector(grid)))

# Function to find neighbors
find_neighbors <- function(current_position) {
  list(
    c(current_position[1] - 1L, current_position[2]),      # Up
    c(current_position[1], current_position[2] + 1L),      # Right
    c(current_position[1] + 1L, current_position[2]),      # Down
    c(current_position[1], current_position[2] - 1L)       # Left
  )
}

# Function to compute complex perimeter conditions
compute_perimeter_complex <- function(coords) {
  if (nrow(coords) < 3) return(4)
  
  coords_complex <- complex(real = coords[, 1], imaginary = coords[, 2])
  directions <- list(-1 + 0i, 1i, 1 + 0i, -1i)
  total_sides <- 0
  
  for (pt in coords_complex) {
    for (i in 1:4) {
      d1 <- directions[[i]]
      d2 <- directions[[i %% 4 + 1]]
      n1 <- pt + d1
      n2 <- pt + d2
      n3 <- pt + d1 + d2
      
      if (n1 %in% coords_complex && n2 %in% coords_complex && !n3 %in% coords_complex) {
        total_sides <- total_sides + 1
      } else if (!n1 %in% coords_complex && !n2 %in% coords_complex) {
        total_sides <- total_sides + 1
      }
    }
  }
  
  total_sides
}

# Function to calculate area and perimeter for a single region
calculate_region_properties <- function(region_char) {
  region_positions <- which(grid == region_char, arr.ind = TRUE)
  position_list <- apply(region_positions, 1, \(x) list(x)[[1]], simplify = FALSE)
  seen <- collections::dict(keys = position_list, items = rep(FALSE, length(position_list)))
  
  total_area_perimeter1 <- 0
  total_area_perimeter2 <- 0
  
  for (i in 1:nrow(region_positions)) {
    current_position <- region_positions[i, ]
    
    if (seen$get(current_position)) next
    seen$set(current_position, TRUE)
    region_touching[] <- -1
    
    # Flood fill to find the entire region
    queue <- collections::queue()
    queue$push(current_position)
    
    while (queue$size() > 0) {
      current <- queue$pop()
      region_touching[current[1], current[2]] <- 0
      
      for (neighbor in find_neighbors(current)) {
        if (neighbor[1] < 1 | neighbor[1] > rows | neighbor[2] < 1 | neighbor[2] > cols) next
        if (grid[neighbor[1], neighbor[2]] != region_char) next
        
        region_touching[current[1], current[2]] <- region_touching[current[1], current[2]] + 1
        
        if (seen$get(neighbor)) next
        
        queue$push(neighbor)
        seen$set(neighbor, TRUE)
      }
    }
    
    # Calculate area and perimeter for the region
    perimeter1 <- sum(4 - region_touching[region_touching != -1])
    area <- sum(region_touching != -1)
    total_area_perimeter1 <- total_area_perimeter1 + area * perimeter1
    
    indices <- which(region_touching != -1, arr.ind = TRUE)
    perimeter2 <- compute_perimeter_complex(indices)
    total_area_perimeter2 <- total_area_perimeter2 + area * perimeter2
  }
  
  c(total_area_perimeter1, total_area_perimeter2)
}

# Initialize results
part1 <- 0
part2 <- 0

# Process each region
for (region_char in region_values) {
  results <- calculate_region_properties(region_char)
  part1 <- part1 + results[1]
  part2 <- part2 + results[2]
}

# Solutions
part1
part2
